pkg/attribute/protoBag.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attribute
+
+import (
+	mixerpb "istio.io/api/mixer/v1"
+)
+
+// ProtoBag implements the Bag interface on top of an Attributes proto.
+type ProtoBag struct {
+	proto          *mixerpb.Attributes
+	globalDict     map[string]int32
+	globalWordList []string
+	messageDict    map[string]int32
+}
+
+func NewProtoBag(proto *mixerpb.Attributes, globalDict map[string]int32, globalWordList []string) *ProtoBag {
+	return &ProtoBag{
+		proto:          proto,
+		globalDict:     globalDict,
+		globalWordList: globalWordList,
+	}
+}
+
+// Get returns an attribute value.
+func (pb *ProtoBag) Get(name string) (interface{}, bool) {
+	index, ok := pb.globalDict[name]
```

Comment:
We might want to look in to a map from index to type, since it's a known mapping for the global dictionary. That way you only do three map lookups: name -> index, index -> type, type[index] -> value. Otherwise in the worst case we have to do 9 lookups. In particular, `request.headers` is frequently used and the last that are hit walking through this method.